### PR TITLE
Correctly handle non handshake commands when using SniHandler

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

As we can only handle handshake commands to parse SNI we should try to skip alert and change cipher spec commands a few times before we fallback to use a default SslContext.

Modifications:

- Use default SslContext if no application data command was received
- Use default SslContext if after 4 commands we not received a handshake command
- Simplify code
- Eliminate multiple volatile fields
- Rename SslConstants to SslUtils
- Share code between SslHandler and SniHandler by moving stuff to SslUtils

Result:

Correct handling of non handshake commands and cleaner code.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin @nmittler @Scottmitch PTAL

**Comment by nmittler** [comment]
@ejona86 might be interested as well

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 1]
```diff
@@ -59,127 +64,161 @@ public SniHandler(DomainNameMapping<? extends SslContext> mapping) {
...
+                                            } else {
+                                                // invalid enum value
+                                                break loop;
+                                            }
+                                        }
+
+                                        offset += extensionLength;
+                                    }
+                                    break loop;
+                                } else {
```
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: I think if you structure this the opposite way, you can get rid of the `break loop` above:

```java
if (in.readableBytes() < packetLength) {
    // client hello incomplete try again to decode once more data is ready.
    return;
}

...
// Fall-through
```

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 2]
```diff
@@ -59,127 +64,161 @@ public SniHandler(DomainNameMapping<? extends SslContext> mapping) {
...
+                                        if (extensionType == 0) {
+                                            int serverNameType = in.getUnsignedByte(offset + 2);
+                                            if (serverNameType == 0) {
+                                                int serverNameLength = in.getUnsignedShort(offset + 3);
+                                                String hostname = in.toString(offset + 5, serverNameLength,
+                                                        CharsetUtil.UTF_8);
+                                                select(ctx, IDN.toASCII(hostname,
+                                                        IDN.ALLOW_UNASSIGNED).toLowerCase(Locale.US));
+                                                return;
+                                            } else {
```
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: nit: don't need the else

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: good point...

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java (hunk 1 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: you are right... let me fix this

**Comment by trustin** [comment]
I wonder creating an extra object to reduce the number of volatile writes seems like an overkill to me.

Besides that, looks good to me. :-)

**Comment by normanmaurer** [comment]
@trustin honestly the object is only created once and it seems more clean (as both fields just be updated together) so I would like to keep it like it is :)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (27118ec53abadac85bc10a3a1aa20220c944058b) and 4.1 (3c5abaa39aa434ac39260164d2e71fda23998bc3)

**Comment by Scottmitch** [comment]
lgtm!

**Pull request closed without merging**